Add categorization to marketplace page

jeff-phillips-18 opened the pull request:
https://jira.coreos.com/browse/CONSOLE-764

Marketplace categories are dynamically created.

![image](https://user-images.githubusercontent.com/11633780/48449874-fc005100-e771-11e8-82fd-616fbb01dc6e.png)

This PR also introduces `category-filter-utils.ts` as utilities for both Catalog and Marketplace for the categorization and filtering of items.

Adds the markdown from the longDescription from the package manifests to the details modal for marketplace items.

Adds URL params to direct link to item details in catalog and marketplace.

@SamiSousa
@tlwu2013
@serenamarie125

(replaces https://github.com/openshift/console/pull/756)

SamiSousa commented:
Looks good to me; great work Jeff!

galletti94 commented on frontend/public/components/app.jsx:
```diff
@@ -145,7 +149,7 @@ class App extends React.PureComponent {
             <LazyRoute path="/start-guide" exact loader={() => import('./start-guide' /* webpackChunkName: "start-guide" */).then(m => m.StartGuidePage)} />
 
-            <LazyRoute path="/marketplace" exact loader={() => import('./marketplace/kubernetes-marketplace' /* webpackChunkName: "marketplace" */).then(m => m.MarketplacePage)} />
+            <LazyOnceRoute path="/marketplace" exact componentName="MarketplacePage" loader={() => import('./marketplace/kubernetes-marketplace' /* webpackChunkName: "marketplace" */)} />
```
Looks like the view of categories for ```/marketplace``` and ```/marketplace?categories=all``` is different. Maybe this should be ```/marketplace?categories=all``` ? Or it should reroute to it somehow? What do you think?

jeff-phillips-18 replied:
There should be no difference. I don't see any. What are you seeing differently?

galletti94 replied:
Sounds good!

galletti94 commented on frontend/public/components/nav.jsx:
```diff
@@ -389,9 +389,7 @@ export class Nav extends React.Component {
-            {
-              // <HrefLink required={FLAGS.KUBERNETES_MARKETPLACE} href="/marketplace" name="Kubernetes Marketplace" activePath="/marketplace/" onClick={this.close} />
-            }
+            <HrefLink required={FLAGS.KUBERNETES_MARKETPLACE} href="/marketplace" name="Kubernetes Marketplace" activePath="/marketplace/" onClick={this.close} />
```
```href="/marketplace?categories=all``` ?

galletti94 commented:
@jeff-phillips-18 I just noticed that the filter counts are off when in the default /marketplace:

![screenshot from 2018-11-15 08-48-36](https://user-images.githubusercontent.com/20077170/48557019-91572e80-e8b3-11e8-9a2b-0c2fb98d23c8.png)
 
Seems like they are all 0. But when I click on all categories and get the list of operators per category the filter counts seem accurate.

jeff-phillips-18 commented:
Can you try again with the latest fix?

spadgett commented on frontend/public/components/catalog/catalog-items.jsx:
```diff
@@ -76,387 +131,305 @@ export class CatalogTileViewPage extends React.Component {
+          No catalog items are being shown due to the filters being applied.
+        </EmptyState.Info>
+        <EmptyState.Help>
+          <button type="text" className="btn btn-link" onClick={() => this.clearFilters()}>Clear All Filters</button>
```
Bind clear filters to this

```ts
clearFilters = () => {
  // ...
};
```

Then you don't need the arrow function here:

```suggestion
          <button type="text" className="btn btn-link" onClick={this.clearFilters}>Clear All Filters</button>
```

spadgett commented on frontend/public/components/marketplace/kubernetes-marketplace-items.jsx:
```diff
@@ -151,62 +165,145 @@ export class MarketplaceTileViewPage extends React.Component {
+  // This function is necessary due to calls to history.replace un-mounting the component before returning
+  updateMountedState(newState) {
+    if (!this.unmounted) {
+      this.setState(newState);
```
We have a `SafetyFirst` class that does something like this, although @alecmerdler is trying to refactor it to be a HOC.

Do we know under what circumstances this happens? I wouldn't expect to need this check unless we are using timeouts, XHRs, resize listeners, etc. (The problem I noticed in @dtaylor113's PR was apparently the one you fixed in the upstream resize listener.)

spadgett commented on frontend/public/components/marketplace/kubernetes-marketplace-items.jsx:
```diff
@@ -151,62 +165,145 @@ export class MarketplaceTileViewPage extends React.Component {
+            placeholder="Filter by keyword..."
+            bsClass="form-control"
+            value={activeFilters.keyword.value}
+            onChange={e => this.onKeywordChange(e.target.value)}
```
Why not make `onKeywordChange` take an event and bind to this, then you can use

```suggestion
            onChange={this.onKeywordChange}
```

spadgett commented on frontend/public/components/marketplace/kubernetes-marketplace-items.jsx:
```diff
@@ -218,66 +315,71 @@ export class MarketplaceTileViewPage extends React.Component {
+          No marketplace items are being shown due to the filters being applied.
+        </EmptyState.Info>
+        <EmptyState.Help>
+          <button type="button" className="btn btn-link" onClick={() => this.clearFilters()}>Clear All Filters</button>
```
Bind `clearFilters` to `this` and use

```suggestion
          <button type="button" className="btn btn-link" onClick={this.clearFilters}>Clear All Filters</button>
```

Arrow functions in render tend can be bad for performance, so it's good to avoid them when not needed.

spadgett commented on frontend/public/components/utils/category-filter-utils.ts:
```diff
@@ -0,0 +1,391 @@
+};
+
+// categorize item under sub and main categories
+const addItem = (item: any, category: any, subcategory: any = null) => {
```
We should be defining types for item, category, and subcategory if we're pulling this stuff out as utilities. We can benefit from type checking here.

spadgett commented on frontend/public/style/_overrides.scss:
```diff
@@ -189,6 +189,9 @@ tags-input .autocomplete .suggestion-item em {
 .catalog-tile-pf-body .catalog-tile-pf-description::after {
   bottom: 15px;
 }
+.capitalize-title .catalog-tile-view-pf-category-header-title {
```
nit: keep classes sorted

spadgett commented on frontend/public/components/utils/category-filter-utils.ts:
```diff
@@ -0,0 +1,391 @@
+  const activeItems = activeCategory ? activeCategory.items : 0;
+  const newFilterCounts = {};
+
+  _.forEach(filterGroups, filterGroup => {
```
style nit: consistently use either `_.forEach` or `_.each`. There's a mix here. We usually use `_.each`

spadgett commented on frontend/public/components/utils/category-filter-utils.ts:
```diff
@@ -0,0 +1,391 @@
+    }
+  });
+
+  if (_.size(activeValues)) {
```
nit: prefer using `_.isEmpty` to check if something is empty

```js
return _.isEmpty(activeValues) ? '' : JSON.stringify(activeValues);
```

spadgett commented on frontend/public/components/utils/category-filter-utils.ts:
```diff
@@ -0,0 +1,391 @@
+};
+
+const getFilterSearchParam = groupFilter => {
+  let activeValues = [];
```
This can be const

spadgett commented on frontend/public/components/utils/category-filter-utils.ts:
```diff
@@ -0,0 +1,391 @@
+    if (groupFilter[typeKey].active) {
+      activeValues.push(typeKey);
+    }
+  });
```
I'd suggest using reduce

```js
const activeValues = _.reduce(groupFilter, (result, ({active}), typeKey) => {
  return active ? result.concat(typeKey) : result;
}, []);
```

jeff-phillips-18 commented on frontend/public/components/utils/category-filter-utils.ts:
```diff
@@ -0,0 +1,391 @@
+  return newFilterCounts;
+};
+
+export const CategoryFilterUtils = {
```
Thanks @spadgett, going to work on this and remove more duplicate code between the two which should help reduce the number of exports.

jeff-phillips-18 commented:
Believe I've addressed all comments. Not bothering to retest until approved or at least close (losing battle).

jeff-phillips-18 commented:
Updated to refactor the CatalogTileViewPage and MarketplaceTileViewPage commonality into new TileViewPage component. Utility functions are now located within the TileViewPage.

Updated tile views to not show subcategories in the view, all tiles belonging to the selected category/subcategory are now shown in the view.

SamiSousa commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -0,0 +1,686 @@
+    if (!_.isEqual(items, prevProps.items)) {
+      const availableFilters = getAvailableFilters(defaultFilters, items, filterGroups);
+      const marketplaceCategories = getAvailableCategories(items);
+      const categories = categorizeItems(items, marketplaceCategories);
```
I think this is missing `itemSorter` as an arg

jeff-phillips-18 replied:
Yup, thanks @SamiSousa.

SamiSousa commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -0,0 +1,686 @@
+      categories: categorizeItems(items, itemsSorter, categories),
+      selectedCategoryId: 'all',
+      availableFilters,
+      activeFilters: availableFilters,
```
I'm a bit confused about the distinction between `activeFilters` and `availableFilters`, could you explain?

jeff-phillips-18 replied:
Available filters are the set of filters that can be used. ActiveFilters are the current state of the filters.

SamiSousa commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -0,0 +1,686 @@
+      const availableFilters = getAvailableFilters(defaultFilters, items, filterGroups);
+      const marketplaceCategories = getAvailableCategories(items);
+      const categories = categorizeItems(items, itemsSorter, marketplaceCategories);
+      this.setState({availableFilters, ...this.getUpdatedState(categories, selectedCategoryId, activeFilters)});
```
Does `activeFilters` get updated to match the potential changes to `availableFilters`?

jeff-phillips-18 replied:
Looking into that @SamiSousa, they ought to be...

jeff-phillips-18 replied:
Updated to correctly update the activeFilters, no longer a need to save availableFilters in state.

spadgett commented:
This isn't new, but we should figure out how to better handle the provider filter when the labels don't match:

![kubernetes marketplace okd 2018-11-19 12-30-21](https://user-images.githubusercontent.com/1167259/48724366-09e12680-ebf7-11e8-943a-c0101ddf9d52.png)

SamiSousa commented:
@tlwu2013 How do we deal with providers being misspelled or using different spellings? Since we have no control over how ISVs set providers in their operators, it may be impractical to enforce a specific spelling for Red Hat, let alone for other providers. 
I see three potential solutions, at least for the short term:
- Hard-code the known different spellings of "Red Hat" and replace them with a single agreed-upon spelling
- Use greatest common substrings to match providers. This could be used to match all "Red Hat" variations as the same provider. (However, other providers had the same shared substring, it would incorrectly group them together.)
- Use a list of predefined providers (such as Red Hat, CoreOS, etc) that operators much match to. If they don't match any of the predefined providers, then they match to "Other" providers.
@serenamarie125  Which should we implement for now?

jeff-phillips-18 commented:
@spadgett Do we have a concept similar to the constants file from origin-web-console? We handled this using settings in that file back in the day.

spadgett commented:
> @spadgett Do we have a concept similar to the constants file from origin-web-console? We handled this using settings in that file back in the day.

We had an extension that let you specify aliases with a few hard-coded ones as defaults, but I'm not in love with that approach.

The best course IMO is to try to enforce consistent metadata, ideally with some review process. We should be able to fix this for Red Hat items at least.

A simple workaround might be to ignore Inc, LLC, etc. from the string for building the filters, but that won't catch everything.

jeff-phillips-18 commented:
@spadgett Added a work around for handling `Inc` and `LLC` differences. More can be added if you can think of any.

jeff-phillips-18 commented on frontend/public/components/marketplace/kubernetes-marketplace-items.jsx:
```diff
@@ -58,227 +16,117 @@ const filterGroups = [
-      if (key === 'keyword') {
-        return counts;
-      }
+  static setURLParams(params) {
```
Fixed

jeff-phillips-18 commented on frontend/public/components/marketplace/kubernetes-marketplace-items.jsx:
```diff
@@ -1,284 +1,174 @@
-      filterCounts: {},
-    };
+  static determineCategories(items) {
+    let newCategories = [];
```
We should add a lint rule for this 😉

spadgett replied:
Good call 👍 

#842

jeff-phillips-18 commented on frontend/public/components/marketplace/kubernetes-marketplace-items.jsx:
```diff
@@ -1,284 +1,174 @@
     }
-    return items;
+
+    return item.name.toLowerCase().includes(filterString) ||
```
https://github.com/openshift/console/issues/845

jeff-phillips-18 commented:
@spadgett Reworked to use a map rather than an array for categories.

jeff-phillips-18 commented:
@alecmerdler SearchPage is not a LazyRoute

jeff-phillips-18 commented on frontend/public/components/app.jsx:
```diff
@@ -89,6 +89,24 @@ const DefaultPage = connectToFlags(FLAGS.OPENSHIFT)(({ flags }) => {
+
+  const saveLoader = () => {
+    return loader().then(m => {
+      cachedRouteComponents[componentName] = m[componentName];
```
Can you explain please?

jeff-phillips-18 replied:
Awesome, thanks!

jeff-phillips-18 replied:
Removed all changes to app.jsx

spadgett commented:
Performance tests are failing:

```
1) Performance test : downloads new bundle for Kubernetes marketplace
   Error: Failed expectation
       at Object.<anonymous> /go/src/github.com/openshift/console/frontend/integration-tests/tests/performance.scenario.ts:86:46
       at step /go/src/github.com/openshift/console/frontend/integration-tests/tests/performance.scenario.ts:32:23
       at Object.next /go/src/github.com/openshift/console/frontend/integration-tests/tests/performance.scenario.ts:13:53
       at fulfilled /go/src/github.com/openshift/console/frontend/integration-tests/tests/performance.scenario.ts:4:58
       at process._tickCallback internal/process/next_tick.js:68:7
```

Screnshot:

![b49bbc6ca9651b664068c3ac36437b3f](https://user-images.githubusercontent.com/1167259/49089183-f81c0680-f228-11e8-910a-48eb06c93b32.png)

alecmerdler commented:
The reason the performance tests are failing is because the bundle size for the `/marketplace` route was explicitly restricted to 100KB more than it was in `master`. This is to call out when the size increases, because its already a huge chunk. The solution is either modify the test to increase the limit, or decrease the JS used (probably due to `patternfly-react`).

jeff-phillips-18 commented:
I am looking into patternfly-react packages to see how to better tree shake and remove unused dependencies. Meanwhile, I can update the limit and file an issue to comeback to fix it once I get a better handle on it. Sound reasonable? @alecmerdler @spadgett

spadgett commented:
> I am looking into patternfly-react packages to see how to better tree shake and remove unused dependencies. Meanwhile, I can update the limit and file an issue to comeback to fix it once I get a better handle on it. Sound reasonable? @alecmerdler @spadgett

Yeah, let's bump the limit for now, but open an issue to investigate for the upcoming sprint. Too many follow on PRs are blocked on this PR, and it's not a new problem. I would like to look at it, though.

jeff-phillips-18 commented:
Added #869

jeff-phillips-18 commented:
Bumped the limits for now. Addressed all other comments and build is passing!

alecmerdler commented on frontend/public/components/catalog/catalog-items.jsx:
```diff
@@ -1,464 +1,204 @@
-import {history} from '../utils';
+import {TileViewPage} from '../utils/tile-view-page';
+
+export const catalogCategories = {
```
> Categories are dynamically created.

Doesn't look like it?

jeff-phillips-18 replied:
Yeah, bad comment. Will fix, thanks.

jeff-phillips-18 replied:
That comment is for `Marketplace` categories which is the addition in the PR (catalog was refactored to use common code with marketplace).

jeff-phillips-18 replied:
Updated the comment to be more specific

alecmerdler commented on frontend/public/components/utils/tile-view-page.jsx:
```diff
@@ -0,0 +1,693 @@
+import * as React from 'react';
+import * as _ from 'lodash-es';
+import * as PropTypes from 'prop-types';
+import CatalogTileView from 'patternfly-react-extensions/dist/js/components/CatalogTileView/CatalogTileView';
```
Tests pass using `import { CatalogTileView } from 'patternfly-react-extensions'`.

jeff-phillips-18 replied:
This will likely pull in a lot more code than required and increase the chunk size.

jeff-phillips-18 replied:
Pulling in index.js at the top level references all components so they won't get shaked. The suggested use is to include only what is used rather than pulling from the top level.

jeff-phillips-18 replied:
OK, I'm mistaken. I made the change and the size of the chunk actually decreased 😕

alecmerdler replied:
Webpack is smart enough to automatically tree-shake if using ES6 module syntax!

jeff-phillips-18 commented:
Thanks @alecmerdler, made the suggested updates.

spadgett commented:
jenkins rebuild

spadgett commented:
/lgtm

spadgett commented:
/hold cancel
